server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java

```diff
@@ -0,0 +1,218 @@
+                timesMasterHasGoneNull++;
+            }
+        }
+        return timesMasterHasGoneNull >= n;
+    }
+
+    private static Set<DiscoveryNode> getDistinctMastersSeen(List<DiscoveryNode> masters) {
+        return masters.stream().filter(Objects::nonNull).collect(Collectors.toSet());
+    }
+
+    /**
+     * Returns the set of distinct non-null master nodes seen in this history.
+     * @return The set of all non-null master nodes seen. Could be empty
+     */
+    public Set<DiscoveryNode> getDistinctMastersSeen() {
+        List<TimeAndMaster> masterHistoryCopy = getMasterHistoryForLast30Minutes(masterHistory);
+        return masterHistoryCopy.stream().map(timeAndMaster -> timeAndMaster.master).filter(Objects::nonNull).collect(Collectors.toSet());
+    }
+
+    /**
+     * Returns true if a non-null master was seen at any point in the last n seconds, or if the last-seen master was more than n seconds
+     * ago and non-null.
+     * @param n The number of seconds to look back
+     * @return true if the current master is non-null or if a non-null master was seen in the last n seconds
+     */
+    public boolean hasSeenMasterInLastNSeconds(int n) {
+        List<TimeAndMaster> masterHistoryCopy = getMasterHistoryForLast30Minutes(masterHistory);
+        long now = nowSupplier.get();
+        TimeValue nSeconds = new TimeValue(n, TimeUnit.SECONDS);
+        long nSecondsAgo = now - nSeconds.getMillis();
+        return getCurrentMaster() != null
+            || masterHistoryCopy.stream().anyMatch(timeAndMaster -> timeAndMaster.time > nSecondsAgo && timeAndMaster.master != null);
+    }
+
+    /*
+     * This method creates a copy of masterHistory that only has entries from more than 30 minutes before now (but leaves the newest
+     * entry in even if it is more than 30 minutes old).
+     */
+    private List<TimeAndMaster> getMasterHistoryForLast30Minutes(List<TimeAndMaster> history) {
+        if (history.size() < 2) {
+            return history;
+        }
+        long now = nowSupplier.get();
+        TimeValue thirtyMinutes = new TimeValue(30, TimeUnit.MINUTES);
+        long thirtyMinutesAgo = now - thirtyMinutes.getMillis();
+        TimeAndMaster mostRecent = history.isEmpty() ? null : history.get(history.size() - 1);
+        history = history.stream().filter(timeAndMaster -> timeAndMaster.time > thirtyMinutesAgo).collect(Collectors.toList());
+        if (history.isEmpty() && mostRecent != null) { // The most recent entry was more than 30 minutes ago
+            history.add(mostRecent);
+        }
+        return history;
+    }
+
+    /**
+     * Clears out anything from masterHistory that is from more than 30 minutes before now (but leaves the newest entry in even if it is
+     * more than 30 minutes old). Rather than being scheduled, this method is called whenever the cluster state changes.
+     */
+    private void removeOldMasterHistory(List<TimeAndMaster> newMasterHistory) {
```

Comment:
Yeah I can do that. I was on the fence about this one because it's only used from one place that creates and mutates that list just above where this is called, and I'd considered just inlining it to avoid concerns like this.